File: cappy/cappy/src/domain/match.ts
import type { Iso, Listing, Match, Owner, Requirement, World } from './types.ts'
import { rating, reliability } from './types.ts'
import { assessFeasibility } from './feasibility.ts'
import { earliestOffer } from './availability.ts'
import { hoursFor, quoteFor } from './pricing.ts'
import { durationLabel } from './categories.ts'

const EARTH_KM = 6371
const toRad = (d: number) => (d * Math.PI) / 180

export function distanceKm(
  a: { lat: number; lng: number },
  b: { lat: number; lng: number },
): number {
  const dLat = toRad(b.lat - a.lat)
  const dLng = toRad(b.lng - a.lng)
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLng / 2) ** 2
  return 2 * EARTH_KM * Math.asin(Math.sqrt(h))
}

/** Min-max to 0..1. All-equal collapses to 0.5 rather than dividing by zero. */
function normalise(values: number[]): number[] {
  const lo = Math.min(...values)
  const hi = Math.max(...values)
  if (hi === lo) return values.map(() => 0.5)
  return values.map((v) => (v - lo) / (hi - lo))
}

// Hand-tuned weights. A learned ranker only makes sense once there is real outcome
// data to learn from, which is what the rating loop is accumulating.
const W = { price: 0.3, soon: 0.2, trust: 0.3, near: 0.2 }

/** Reliability dominates; stars break ties; an unrated owner sits mid-scale. */
export function trustOf(o: Owner): number {
  const stars = rating(o)
  return 0.6 * reliability(o) + 0.4 * (stars === null ? 0.5 : stars / 5)
}

export function trackRecord(o: Owner): string {
  if (o.jobsDone === 0) return 'New on Cappy'
  const pct = Math.round((o.onTimeJobs / o.jobsDone) * 100)
  return `${o.jobsDone} booking${o.jobsDone === 1 ? '' : 's'} · ${pct}% on time`
}

const windowEnd = (req: Requirement): Iso =>
  req.mode === 'window' ? req.latest : req.deadline

/**
 * Pure. Every input arrives as an argument, including `now`, so the same inputs
 * always produce the same ranking, which is what makes it testable and portable.
 */
export function findMatches(req: Requirement, world: World, now: Iso): Match[] {
  const from = world.districts[req.district]
  if (!from) return []

  const owners = new Map(world.owners.map((o) => [o.id, o]))
  const until = windowEnd(req)
  if (Date.parse(until) <= Date.parse(now)) return []

  type Candidate = { listing: Listing; owner: Owner; match: Omit<Match, 'score'> }
  const candidates: Candidate[] = []

  for (const listing of world.listings) {
    const owner = owners.get(listing.ownerId)
    if (!owner) continue

    const fit = assessFeasibility(req, listing)
    if (!fit.feasible) continue

    const to = world.districts[listing.district]
    if (!to) continue
    const km = distanceKm(from, to)
    if (km > req.maxDistanceKm) continue

    const hours = hoursFor(req, listing)
    const quote = quoteFor(req, listing)
    if (hours === null || quote === null) continue

    const slots = world.slots.filter((s) => s.listingId === listing.id)
    const offer = earliestOffer(slots, hours, now, until)
    if (!offer) continue

    candidates.push({
      listing,
      owner,
      match: {
        listingId: listing.id,
        ownerId: owner.id,
        slotId: offer.slotId,
        start: offer.start,
        end: offer.end,
        confidence: fit.confidence,
        reasons: [
          ...fit.reasons,
          `${durationLabel(hours)} of idle time`,
          `${km.toFixed(1)} km away`,
          trackRecord(owner),
        ],
        quote,
        distanceKm: km,
      },
    })
  }

  if (candidates.length === 0) return []

  const cheap = normalise(candidates.map((c) => -c.match.quote.total))
  const soon = normalise(candidates.map((c) => -Date.parse(c.match.start)))
  const near = normalise(candidates.map((c) => -c.match.distanceKm))
  const trust = normalise(candidates.map((c) => trustOf(c.owner)))

  return candidates
    .map((c, i) => ({
      ...c.match,
      score: W.price * cheap[i] + W.soon * soon[i] + W.near * near[i] + W.trust * trust[i],
    }))
    .sort((a, b) => b.score - a.score)
}

export type SortKey = 'best' | 'price' | 'soonest' | 'nearest'

export function sortMatches(matches: Match[], key: SortKey): Match[] {
  const copy = [...matches]
  switch (key) {
    case 'price':
      return copy.sort((a, b) => a.quote.total - b.quote.total)
    case 'soonest':
      return copy.sort((a, b) => Date.parse(a.start) - Date.parse(b.start))
    case 'nearest':
      return copy.sort((a, b) => a.distanceKm - b.distanceKm)
    default:
      return copy.sort((a, b) => b.score - a.score)
  }
}

/**
 * Build the match for a window the buyer picked themselves, rather than the
 * soonest one we suggested. Same shape, same reasons, so everything downstream
 * (quote, booking, receipt) is identical whichever route they took.
 */
export function matchForOffer(
  req: Requirement,
  listing: Listing,
  owner: Owner,
  offer: { slotId: string; start: Iso; end: Iso },
  km: number,
): Match | null {
  const fit = assessFeasibility(req, listing)
  if (!fit.feasible) return null
  const hours = hoursFor(req, listing)
  const quote = quoteFor(req, listing)
  if (hours === null || quote === null) return null

  return {
    listingId: listing.id,
    ownerId: owner.id,
    slotId: offer.slotId,
    start: offer.start,
    end: offer.end,
    score: 1,
    confidence: fit.confidence,
    reasons: [
      ...fit.reasons,
      `${durationLabel(hours)} of idle time`,
      `${km.toFixed(1)} km away`,
      trackRecord(owner),
    ],
    quote,
    distanceKm: km,
  }
}

File: backend/services/matching/matching/domain/match.py
"""Port of ``src/domain/match.ts``."""

from __future__ import annotations

import math
from typing import Literal

from cappy_common.jsmath import js_round, to_fixed
from cappy_common.models import (
    AnyListing,
    AnyRequirement,
    Iso,
    Match,
    Owner,
    Quote,
    World,
    rating,
    reliability,
)
from cappy_common.timeutil import ms_from_iso

from .availability import Offer, earliest_offer
from .categories import duration_label
from .feasibility import assess_feasibility
from .pricing import hours_for, quote_for

EARTH_KM = 6371


def distance_km(a: tuple[float, float], b: tuple[float, float]) -> float:
    """Haversine between two (lat, lng) pairs."""
    lat1, lng1 = a
    lat2, lng2 = b
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    h = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lng / 2) ** 2
    )
    return 2 * EARTH_KM * math.asin(math.sqrt(h))


def _normalise(values: list[float]) -> list[float]:
    """Min-max to 0..1. All-equal collapses to 0.5 rather than dividing by zero."""
    lo, hi = min(values), max(values)
    if hi == lo:
        return [0.5 for _ in values]
    return [(v - lo) / (hi - lo) for v in values]


# Hand-tuned weights. A learned ranker only makes sense once there is real outcome
# data to learn from, which is what the rating loop is accumulating.
W = {"price": 0.3, "soon": 0.2, "trust": 0.3, "near": 0.2}


def trust_of(o: Owner) -> float:
    """Reliability dominates; stars break ties; an unrated owner sits mid-scale."""
    stars = rating(o)
    return 0.6 * reliability(o) + 0.4 * (0.5 if stars is None else stars / 5)


def track_record(o: Owner) -> str:
    if o.jobs_done == 0:
        return "New on Cappy"
    pct = js_round((o.on_time_jobs / o.jobs_done) * 100)
    return f"{o.jobs_done} booking{'' if o.jobs_done == 1 else 's'} · {pct}% on time"


def _window_end(req: AnyRequirement) -> Iso:
    return req.latest if req.mode == "window" else req.deadline


def _reasons(fit_reasons: list[str], hours: float, km: float, owner: Owner) -> list[str]:
    return [
        *fit_reasons,
        f"{duration_label(hours)} of idle time",
        f"{to_fixed(km, 1)} km away",
        track_record(owner),
    ]


def find_matches(req: AnyRequirement, world: World, now: Iso) -> list[Match]:
    """Pure. Every input arrives as an argument, including ``now``."""
    origin = world.districts.get(req.district)
    if not origin:
        return []

    owners = {o.id: o for o in world.owners}
    until = _window_end(req)
    if ms_from_iso(until) <= ms_from_iso(now):
        return []

    slots_by_listing: dict[str, list] = {}
    for s in world.slots:
        slots_by_listing.setdefault(s.listing_id, []).append(s)

    candidates: list[tuple[Owner, dict]] = []

    for listing in world.listings:
        owner = owners.get(listing.owner_id)
        if not owner:
            continue

        fit = assess_feasibility(req, listing)
        if not fit.feasible:
            continue

        dest = world.districts.get(listing.district)
        if not dest:
            continue
        km = distance_km((origin.lat, origin.lng), (dest.lat, dest.lng))
        if km > req.max_distance_km:
            continue

        hours = hours_for(req, listing)
        quote = quote_for(req, listing)
        if hours is None or quote is None:
            continue

        offer = earliest_offer(slots_by_listing.get(listing.id, []), hours, now, until)
        if not offer:
            continue

        candidates.append(
            (
                owner,
                dict(
                    listing_id=listing.id,
                    owner_id=owner.id,
                    slot_id=offer.slot_id,
                    start=offer.start,
                    end=offer.end,
                    confidence=fit.confidence,
                    reasons=_reasons(fit.reasons, hours, km, owner),
                    quote=quote,
                    distance_km=km,
                ),
            )
        )

    if not candidates:
        return []

    cheap = _normalise([-c["quote"].total for _, c in candidates])
    soon = _normalise([-ms_from_iso(c["start"]) for _, c in candidates])
    near = _normalise([-c["distance_km"] for _, c in candidates])
    trust = _normalise([trust_of(o) for o, _ in candidates])

    matches = [
        Match(
            **c,
            score=W["price"] * cheap[i] + W["soon"] * soon[i] + W["near"] * near[i] + W["trust"] * trust[i],
        )
        for i, (_, c) in enumerate(candidates)
    ]
    return sorted(matches, key=lambda m: -m.score)


SortKey = Literal["best", "price", "soonest", "nearest"]


def sort_matches(matches: list[Match], key: SortKey) -> list[Match]:
    if key == "price":
        return sorted(matches, key=lambda m: m.quote.total)
    if key == "soonest":
        return sorted(matches, key=lambda m: ms_from_iso(m.start))
    if key == "nearest":
        return sorted(matches, key=lambda m: m.distance_km)
    return sorted(matches, key=lambda m: -m.score)


def match_for_offer(req: AnyRequirement, listing: AnyListing, owner: Owner, offer: Offer, km: float) -> Match | None:
    """Build the match for a window the buyer picked themselves, rather than the
    soonest one we suggested. Same shape, same reasons, so everything downstream
    (quote, booking, receipt) is identical whichever route they took."""
    fit = assess_feasibility(req, listing)
    if not fit.feasible:
        return None
    hours = hours_for(req, listing)
    quote: Quote | None = quote_for(req, listing)
    if hours is None or quote is None:
        return None

    return Match(
        listing_id=listing.id,
        owner_id=owner.id,
        slot_id=offer.slot_id,
        start=offer.start,
        end=offer.end,
        score=1,
        confidence=fit.confidence,
        reasons=_reasons(fit.reasons, hours, km, owner),
        quote=quote,
        distance_km=km,
    )
